colossalai/gemini/memory_tracer/memstats_collector.py

```diff
@@ -171,54 +172,86 @@ def __init__(self, module: nn.Module, chunk_manager: ChunkManager) -> None:
 
-    def init_mem_stats(self, *inputs):
+    def init_mem_stats(self, **kwargs):
 
         self.register_opnodes_recursively(self.module)
         self.refactor_module()
 
-        self.module = self.module.cpu()
+        # self.module = self.module.cpu()
         self.module.train()
 
-        data = [MetaTensor(torch.rand(inp.shape, device='meta'), fake_device='cpu') for inp in inputs]
-        gm = symbolic_trace(self.module)
+        graph = ColoTracer().trace(self.module, meta_args=kwargs)
+        gm = torch.fx.GraphModule(self.module, graph)
         interp = MetaInfoProp(gm)
-        interp.propagate(*data)
+        interp.propagate(*[MetaTensor(v, fake_device='cuda:0') for k, v in kwargs.items()])
 
-        total_mem = 0
-        for inp in inputs:
-            total_mem += inp.numel() * inp.element_size()
-        last_node = None
         module_name_list = [mInfo.module_full_name for mInfo in self.module_info_list]
+        fwd_out_released = {}
+        total_mem = 0
+
         for node in gm.graph.nodes:
             total_mem = total_mem + calculate_fwd_tmp(node) + calculate_fwd_out(node)
+            if calculate_fwd_out(node) > 0:
+                fwd_out_released[node] = False
             if node.op == "call_module":
                 if node.name.endswith("_0") and node.name[:-2] in module_name_list:
                     self._non_model_data_cuda_list.append(total_mem)
-                last_node = node
+
         self._non_model_data_cuda_list.append(total_mem)
         self._non_model_data_cuda_list = self._non_model_data_cuda_list[1:]
 
-        cur_module_mem_fwd = 0
-        cur_module_mem_bwd = 0
-        grad_module_out = last_node.meta["fwd_mem_out"]
+        peak_mem = total_mem
+        grad_in_computed = {}
+
         for node in gm.graph.nodes.__reversed__():
-            cur_module_mem_fwd = cur_module_mem_fwd + calculate_fwd_tmp(node) + calculate_fwd_out(node)
-            cur_module_mem_bwd = cur_module_mem_bwd + node.meta["bwd_mem_tmp"] + node.meta["bwd_mem_out"]
+
+            if node.name.__contains__("where") or node.name.__contains__("truediv"):
+                continue
+
+            # before run backward of the node
+
+            total_mem = total_mem + node.meta["bwd_mem_tmp"] + node.meta["bwd_mem_out"]
+            if total_mem >= peak_mem:
+                peak_mem = total_mem
+
+            # after run backward of the node
+
+            total_mem -= node.meta["bwd_mem_tmp"]
+            total_mem -= calculate_fwd_tmp(node)
+
+            # release grad_in of current node
+            for grad_in in node.meta["fwd_out"]:
+                if isinstance(grad_in, torch.Tensor):
+                    total_mem -= grad_in.numel() * 4.0
+
+            for in_node in node.args:
+                if isinstance(in_node, torch.fx.node.Node):
+                    # release fwd_in (fwd_out) of current node (input nodes)
+                    if calculate_fwd_out(in_node) > 0 and (not fwd_out_released[in_node]):
```

Comment:
Does these lines deal with the checkpoint case? If so, please add some comments to make it clear.